openrtb-json/src/main/scala/com/powerspace/openrtb/json/bidrequest/OpenRtbNativeRequestSerde.scala

```diff
@@ -1,48 +1,88 @@
   }
 
-  private implicit val assetEncoder: Encoder[NativeRequest.Asset] = openrtbEncoder[Asset]
-  private implicit val eventTrackersEncoder: Encoder[NativeRequest.EventTrackers] = openrtbEncoder[EventTrackers]
+  private implicit val assetEncoder: Encoder[NativeRequest.Asset] = openRtbEncoder[Asset]
+  private implicit val eventTrackersEncoder: Encoder[NativeRequest.EventTrackers] = openRtbEncoder[EventTrackers]
+  implicit val nativeRequestEncoder: Encoder[NativeRequest] = openRtbEncoder[NativeRequest]
 
   private implicit val requestOneOfEncoder: Encoder[Imp.Native.RequestOneof] = protobufOneofEncoder[Imp.Native.RequestOneof] {
     case Imp.Native.RequestOneof.Request(string) => string.asJson
     case Imp.Native.RequestOneof.RequestNative(request) => request.asJson
   }
 
-  implicit val nativeRequestEncoder: Encoder[NativeRequest] = openrtbEncoder[NativeRequest]
+  def encoder: Encoder[Imp.Native] = openRtbEncoder[Imp.Native]
+
+  private implicit val titleDecoder: Decoder[NativeRequest.Asset.Title] = openRtbDecoder[Asset.Title]
+  private implicit val imgDecoder: Decoder[NativeRequest.Asset.Image] = openRtbDecoder[Asset.Image]
+  private implicit val videoDecoder: Decoder[Imp.Video] = OpenRtbVideoSerde.decoder
+  private implicit val assetDataDecoder: Decoder[NativeRequest.Asset.Data] = openRtbDecoder[Asset.Data]
+
+  private implicit val assetDecoder: Decoder[NativeRequest.Asset] = cursor =>
+    for {
+      id <- cursor.downField("id").as[Int]
+      required <- cursor.downField("required").as[Option[Boolean]]
+      title <- cursor.downField("title").as[Option[NativeRequest.Asset.Title]].map(_.map(NativeRequest.Asset.AssetOneof.Title))
+      img <- cursor.downField("img").as[Option[NativeRequest.Asset.Image]].map(_.map(NativeRequest.Asset.AssetOneof.Img))
+      video <- cursor.downField("video").as[Option[Imp.Video]].map(_.map(NativeRequest.Asset.AssetOneof.Video))
+      data <- cursor.downField("data").as[Option[NativeRequest.Asset.Data]].map(_.map(NativeRequest.Asset.AssetOneof.Data))
+      assetOneof = title
+        .orElse(img)
+        .orElse(video)
+        .orElse(data)
+        .getOrElse(AssetOneof.Empty)
+    } yield {
+      NativeRequest.Asset(id = id, required = required, assetOneof = assetOneof)
+    }
+
+  private implicit val eventTrackersDecoder: Decoder[NativeRequest.EventTrackers] = openRtbDecoder[EventTrackers]
+  implicit val nativeRequestDecoder: Decoder[NativeRequest] = openRtbDecoder[NativeRequest]
+
+  private implicit val requestOneOfDecoder: Decoder[Imp.Native.RequestOneof] = (cursor: HCursor) =>
+    cursor.focus.map {
+      case json if json.isString => json.asString.map(RequestOneof.Request).getOrElse(RequestOneof.Empty)
+      case json if json.isObject => json.as[NativeRequest].map(RequestOneof.RequestNative).getOrElse(RequestOneof.Empty)
+    }.orElse(Some(RequestOneof.Empty)).map(Right(_)).get
+
+  def decoder: Decoder[Imp.Native] = cursor =>
```

Comment:
Pretty sure this can be semi auto

Reply:
It can't because circa doesn't bind the `request` field with the `requestOneof` decoder. But I could manage to shrink anyways :)